torture/src/supervisor/workload.rs

```diff
@@ -414,43 +414,107 @@ impl Workload {
             }
         }
 
-        if self.state.rng.gen_bool(self.state.biases.commit_crash) {
-            self.exercise_commit_crashing(&rr).await?;
-        } else {
-            self.exercise_commit(&rr).await?;
-        }
+        let should_crash = self.state.rng.gen_bool(self.state.biases.commit_crash);
+        self.exercise_commit(&rr, should_crash).await?;
 
         Ok(())
     }
 
     /// Commit a changeset.
-    async fn exercise_commit(&mut self, rr: &comms::RequestResponse) -> anyhow::Result<()> {
-        trace!("exercising commit");
+    async fn exercise_commit(
+        &mut self,
+        rr: &comms::RequestResponse,
+        should_crash: bool,
+    ) -> anyhow::Result<()> {
+        let should_crash = if should_crash {
+            trace!("exercising commit crash");
+            Some(self.get_crash_delay())
+        } else {
+            trace!("exercising commit");
+            None
+        };
 
+        // Generate a changeset and the associated snapshot
         let (snapshot, changeset) = self.state.gen_commit();
-        let ToSupervisor::CommitResponse { elapsed, outcome } = rr
+        let commit_response = rr
             .send_request(crate::message::ToAgent::Commit(
                 crate::message::CommitPayload {
                     changeset: changeset.clone(),
-                    should_crash: None,
+                    should_crash,
                 },
             ))
-            .await?
-        else {
-            return Err(anyhow::anyhow!("Commit did not execute successfully"));
+            .await?;
+
+        // Ensure that the outcome is what we expect.
+        if should_crash.is_some() {
+            self.wait_for_crash().await?;
+            let ToSupervisor::Ack = commit_response else {
+                return Err(anyhow::anyhow!("Commit crash did not execute successfully"));
+            };
+
+            // During a commit crash, every type of error could happen.
+            // However the agent will be respawned, so it will just
+            // make sure the changeset was correctly applied or reverted.
+            self.spawn_new_agent().await?;
+        } else {
+            let ToSupervisor::CommitResponse { elapsed, outcome } = commit_response else {
+                return Err(anyhow::anyhow!("Commit did not execute successfully"));
+            };
+
+            self.ensure_outcome_validity(rr, &outcome).await?;
+
+            if matches!(outcome, crate::message::Outcome::Success) {
+                self.n_successfull_commit += 1;
+                self.tot_commit_time += elapsed;
+            }
         };
 
+        // Udpate RequestResponse, the agent could have been respawned.
```

Comment:
```suggestion
        // Update RequestResponse, the agent could have been respawned.
```

although this particular action is often referred as "reload", so a better version would be

```suggestion
        // Reload RequestResponse, the agent could have been respawned.
```